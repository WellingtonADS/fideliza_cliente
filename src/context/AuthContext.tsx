import React, { createContext, useState, useEffect, useContext, ReactNode } from 'react';
import AsyncStorage from '@react-native-async-storage/async-storage';
import { User, UserCredentials, UserRegistration } from '../types/auth';
import api, { login as apiLogin, registerClient, getMyProfile, setAuthToken } from '../services/api';
import axios from 'axios';
import Toast from 'react-native-toast-message';

interface AuthContextType {
  token: string | null;
  user: User | null;
  isLoading: boolean;
  signIn: (credentials: UserCredentials) => Promise<void>;
  signUp: (userData: UserRegistration) => Promise<void>;
  signOut: () => void;
  refreshUser: () => Promise<void>;
}

// CORREÇÃO: Adicionado 'export' para que o contexto possa ser importado em outros lugares
export const AuthContext = createContext<AuthContextType | undefined>(undefined);

export const AuthProvider = ({ children }: { children: ReactNode }) => {
  const [token, setToken] = useState<string | null>(null);
  const [user, setUser] = useState<User | null>(null);
  const [isLoading, setIsLoading] = useState(true);

  useEffect(() => {
    const loadTokenFromStorage = async () => {
      try {
        const storedToken = await AsyncStorage.getItem('userToken');
        if (storedToken) {
          setToken(storedToken);
          setAuthToken(storedToken);
          const response = await getMyProfile();
          setUser(response.data);
        }
      } catch (error) {
        console.error("Falha ao carregar token do storage", error);
        await signOut(); 
      } finally {
        setIsLoading(false);
      }
    };
    loadTokenFromStorage();
  }, []);

  

  const signIn = async (credentials: UserCredentials) => {
    try {
      const response = await apiLogin(credentials);
      const { access_token } = response.data;
      setToken(access_token);
      setAuthToken(access_token);
      await AsyncStorage.setItem('userToken', access_token);
      const profileResponse = await getMyProfile();
      setUser(profileResponse.data);
    } catch (error) {
      if (axios.isAxiosError(error) && error.response) {
        if (error.response.status === 401) {
          throw new Error('Email ou senha inválidos.');
        }
      }
      throw new Error('Não foi possível conectar ao servidor. Tente novamente.');
    }
  };

  const signUp = async (userData: UserRegistration) => {
    try {
      await registerClient(userData);
      await signIn({ email: userData.email, password: userData.password });
    } catch (error) {
      if (axios.isAxiosError(error) && error.response?.data?.detail) {
         throw new Error(error.response.data.detail);
      }
      throw new Error('Ocorreu um erro ao tentar registar. Tente novamente.');
    }
  };

  const signOut = async () => {
    setToken(null);
    setUser(null);
    setAuthToken(undefined);
    await AsyncStorage.removeItem('userToken');
  };

  const refreshUser = async () => {
    if (!token) return;
    try {
      const profileResponse = await getMyProfile();
      setUser(profileResponse.data);
    } catch (error) {
      await signOut();
    }
  };

  // Interceptor global para 401 -> sessão expirada (precisa de signOut definido)
  useEffect(() => {
    const interceptorId = api.interceptors.response.use(
      (response) => response,
      async (error) => {
        if (axios.isAxiosError(error) && error.response) {
          if (error.response.status === 401) {
          // Evita disparar em endpoints de login
          const url = error.config?.url || '';
          if (!url.includes('/token')) {
            Toast.show({ type: 'error', text1: 'Sessão expirada', text2: 'Faça login novamente.' });
            await signOut();
          }
          } else if (error.response.status === 403) {
            Toast.show({ type: 'error', text1: 'Acesso negado', text2: 'Você não tem permissão para esta ação.' });
          }
        }
        return Promise.reject(error);
      }
    );
    return () => {
      api.interceptors.response.eject(interceptorId);
    };
  }, [signOut]);

  return (
    <AuthContext.Provider value={{ token, user, isLoading, signIn, signUp, signOut, refreshUser }}>
      {children}
    </AuthContext.Provider>
  );
};

export const useAuth = (): AuthContextType => {
  const context = useContext(AuthContext);
  if (context === undefined) {
    throw new Error('useAuth deve ser usado dentro de um AuthProvider');
  }
  return context;
};
