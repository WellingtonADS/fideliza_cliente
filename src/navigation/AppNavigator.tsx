// src/navigation/AppNavigator.tsx
import React from 'react';
import { View, ActivityIndicator, StyleSheet, Text } from 'react-native'; // Adicionado Text
import { colors } from '../theme/colors';
import { useAuth } from '../context/AuthContext';

import AuthNavigator from './AuthNavigator';
import MainNavigator from './MainNavigator'; 


const AppNavigator = () => {
  const { token, isLoading } = useAuth();

  if (isLoading) {
    return (
      <View style={styles.container}>
        <ActivityIndicator size="large" color="#FFFFFF" />
        <Text style={styles.loadingText}>A carregar...</Text>
      </View>
    );
  }

  return token ? <MainNavigator /> : <AuthNavigator />;
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: colors.background, // Cor de fundo escura, consistente com o resto do app
  },
  loadingText: {
    marginTop: 10,
    color: colors.text,
    fontSize: 16,
  }
});

export default AppNavigator;
