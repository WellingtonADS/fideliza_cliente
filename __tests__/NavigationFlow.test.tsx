// Arquivo LIMPO: teste reduzido e estável de fluxo de navegação
import React from 'react';
import TestRenderer from 'react-test-renderer';
import { NavigationContainer } from '@react-navigation/native';
import AppNavigator from '../src/navigation/AppNavigator';
import { AuthContext } from '../src/context/AuthContext';

jest.mock('../src/services/api'); // usa __mocks__/src/services/api.ts

interface MockAuthValue {
  token: string | null;
  user: any | null;
  isLoading: boolean;
  signIn: jest.Mock;
  signUp: jest.Mock;
  signOut: jest.Mock;
  refreshUser: jest.Mock;
}

function sanitize(node: any): any {
  if (node == null) return node;
  if (typeof node === 'string' || typeof node === 'number') return node;
  if (Array.isArray(node)) return node.map(sanitize);
  if (typeof node === 'object') {
    const { props, children, ...rest } = node as any;
    const newProps = { ...(props || {}) };
    if ('screenId' in newProps) delete newProps.screenId;
    return { ...rest, props: newProps, children: sanitize(children) };
  }
  return node;
}

async function renderWith(value: MockAuthValue) {
  let tree: TestRenderer.ReactTestRenderer | undefined;
  await TestRenderer.act(async () => {
    tree = TestRenderer.create(
      <AuthContext.Provider value={value as any}>
        <NavigationContainer>
          <AppNavigator />
        </NavigationContainer>
      </AuthContext.Provider>
    );
  });
  await TestRenderer.act(async () => { await Promise.resolve(); });
  return tree!;
}

describe('NavigationFlow', () => {
  const baseValue: MockAuthValue = {
    token: null,
    user: null,
    isLoading: false,
    signIn: jest.fn(),
    signUp: jest.fn(),
    signOut: jest.fn(),
    refreshUser: jest.fn(),
  };

  it('mostra telas de autenticação quando não logado', async () => {
    const tree = await renderWith(baseValue);
    expect(sanitize(tree.toJSON())).toMatchSnapshot();
  });

  it('mostra fluxo principal quando logado', async () => {
    const tree = await renderWith({
      ...baseValue,
      token: 'token',
      user: { id: 1, name: 'User', email: 'user@test.com' },
    });
    expect(sanitize(tree.toJSON())).toMatchSnapshot();
  });
});
