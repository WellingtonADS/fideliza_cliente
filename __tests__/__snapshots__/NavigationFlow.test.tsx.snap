// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavigationFlow mostra fluxo principal quando logado 1`] = `
{
  "children": [
    {
      "children": null,
      "props": {
        "onInsetsChange": [Function],
        "style": [
          {
            "flex": 1,
          },
          {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          },
        ],
      },
      "type": "RNCSafeAreaProvider",
    },
    {
      "children": [
        {
          "children": [
            {
              "children": "[...]",
              "props": {
                "collapsable": false,
                "style": [
                  {
                    "flex": 1,
                  },
                  [
                    {
                      "backgroundColor": "rgb(242, 242, 242)",
                    },
                    undefined,
                  ],
                ],
              },
              "type": "RNSScreenContentWrapper",
            },
            {
              "children": "[...]",
              "props": {
                "backButtonDisplayMode": undefined,
                "backButtonInCustomView": false,
                "backTitle": undefined,
                "backTitleFontFamily": "System",
                "backTitleFontSize": undefined,
                "backTitleVisible": true,
                "backgroundColor": "rgb(255, 255, 255)",
                "blurEffect": undefined,
                "color": "rgb(0, 122, 255)",
                "direction": "ltr",
                "disableBackButtonMenu": false,
                "hidden": true,
                "hideBackButton": false,
                "hideShadow": undefined,
                "largeTitle": undefined,
                "largeTitleBackgroundColor": undefined,
                "largeTitleColor": undefined,
                "largeTitleFontFamily": "System",
                "largeTitleFontSize": undefined,
                "largeTitleFontWeight": "700",
                "largeTitleHideShadow": false,
                "pointerEvents": "box-none",
                "style": {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                  "position": "absolute",
                  "width": "100%",
                },
                "title": "Home",
                "titleColor": "rgb(28, 28, 30)",
                "titleFontFamily": "System",
                "titleFontSize": undefined,
                "titleFontWeight": "600",
                "topInsetEnabled": true,
                "translucent": false,
              },
              "type": "RNSScreenStackHeaderConfig",
            },
          ],
          "props": {
            "activityState": 2,
            "aria-hidden": false,
            "collapsable": false,
            "customAnimationOnSwipe": undefined,
            "fullScreenSwipeEnabled": undefined,
            "fullScreenSwipeShadowEnabled": true,
            "gestureEnabled": undefined,
            "gestureResponseDistance": {
              "bottom": -1,
              "end": -1,
              "start": -1,
              "top": -1,
            },
            "hasLargeHeader": false,
            "hideKeyboardOnSwipe": undefined,
            "homeIndicatorHidden": undefined,
            "nativeBackButtonDismissalEnabled": false,
            "navigationBarHidden": undefined,
            "onAppear": [Function],
            "onDisappear": [Function],
            "onDismissed": [Function],
            "onGestureCancel": [Function],
            "onHeaderBackButtonClicked": [Function],
            "onHeaderHeightChange": [Function],
            "onNativeDismissCancelled": [Function],
            "onSheetDetentChanged": [Function],
            "onTransitionProgress": [Function],
            "onWillAppear": [Function],
            "onWillDisappear": [Function],
            "preventNativeDismiss": undefined,
            "replaceAnimation": "push",
            "screenOrientation": undefined,
            "sheetAllowedDetents": [
              1,
            ],
            "sheetCornerRadius": -1,
            "sheetElevation": 24,
            "sheetExpandsWhenScrolledToEdge": true,
            "sheetGrabberVisible": false,
            "sheetInitialDetent": 0,
            "sheetInitialDetentIndex": 0,
            "sheetLargestUndimmedDetent": -1,
            "sheetLargestUndimmedDetentIndex": -1,
            "stackAnimation": undefined,
            "stackPresentation": "push",
            "statusBarAnimation": undefined,
            "statusBarHidden": undefined,
            "statusBarStyle": undefined,
            "style": {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
              "zIndex": undefined,
            },
            "swipeDirection": "horizontal",
            "transitionDuration": undefined,
          },
          "type": "RNSScreen",
        },
      ],
      "props": {
        "onFinishTransitioning": undefined,
        "style": {
          "flex": 1,
        },
      },
      "type": "RNSScreenStack",
    },
  ],
  "props": {
    "onInsetsChange": [Function],
    "style": [
      {
        "flex": 1,
      },
      undefined,
    ],
  },
  "type": "RNCSafeAreaProvider",
}
`;

exports[`NavigationFlow mostra fluxo principal quando logado 2`] = `
{
  "children": [
    {
      "children": null,
      "props": {
        "onInsetsChange": [Function],
        "style": [
          {
            "flex": 1,
          },
          {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          },
        ],
      },
      "type": "RNCSafeAreaProvider",
    },
    {
      "children": [
        {
          "children": [
            {
              "children": "[...]",
              "props": {
                "collapsable": false,
                "style": [
                  {
                    "flex": 1,
                  },
                  [
                    {
                      "backgroundColor": "rgb(242, 242, 242)",
                    },
                    undefined,
                  ],
                ],
              },
              "type": "RNSScreenContentWrapper",
            },
            {
              "children": "[...]",
              "props": {
                "backButtonDisplayMode": undefined,
                "backButtonInCustomView": false,
                "backTitle": undefined,
                "backTitleFontFamily": "System",
                "backTitleFontSize": undefined,
                "backTitleVisible": true,
                "backgroundColor": "rgb(255, 255, 255)",
                "blurEffect": undefined,
                "color": "rgb(0, 122, 255)",
                "direction": "ltr",
                "disableBackButtonMenu": false,
                "hidden": true,
                "hideBackButton": false,
                "hideShadow": undefined,
                "largeTitle": undefined,
                "largeTitleBackgroundColor": undefined,
                "largeTitleColor": undefined,
                "largeTitleFontFamily": "System",
                "largeTitleFontSize": undefined,
                "largeTitleFontWeight": "700",
                "largeTitleHideShadow": false,
                "pointerEvents": "box-none",
                "style": {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                  "position": "absolute",
                  "width": "100%",
                },
                "title": "Home",
                "titleColor": "rgb(28, 28, 30)",
                "titleFontFamily": "System",
                "titleFontSize": undefined,
                "titleFontWeight": "600",
                "topInsetEnabled": true,
                "translucent": false,
              },
              "type": "RNSScreenStackHeaderConfig",
            },
          ],
          "props": {
            "activityState": 2,
            "aria-hidden": false,
            "collapsable": false,
            "customAnimationOnSwipe": undefined,
            "fullScreenSwipeEnabled": undefined,
            "fullScreenSwipeShadowEnabled": true,
            "gestureEnabled": undefined,
            "gestureResponseDistance": {
              "bottom": -1,
              "end": -1,
              "start": -1,
              "top": -1,
            },
            "hasLargeHeader": false,
            "hideKeyboardOnSwipe": undefined,
            "homeIndicatorHidden": undefined,
            "nativeBackButtonDismissalEnabled": false,
            "navigationBarHidden": undefined,
            "onAppear": [Function],
            "onDisappear": [Function],
            "onDismissed": [Function],
            "onGestureCancel": [Function],
            "onHeaderBackButtonClicked": [Function],
            "onHeaderHeightChange": [Function],
            "onNativeDismissCancelled": [Function],
            "onSheetDetentChanged": [Function],
            "onTransitionProgress": [Function],
            "onWillAppear": [Function],
            "onWillDisappear": [Function],
            "preventNativeDismiss": undefined,
            "replaceAnimation": "push",
            "screenOrientation": undefined,
            "sheetAllowedDetents": [
              1,
            ],
            "sheetCornerRadius": -1,
            "sheetElevation": 24,
            "sheetExpandsWhenScrolledToEdge": true,
            "sheetGrabberVisible": false,
            "sheetInitialDetent": 0,
            "sheetInitialDetentIndex": 0,
            "sheetLargestUndimmedDetent": -1,
            "sheetLargestUndimmedDetentIndex": -1,
            "stackAnimation": undefined,
            "stackPresentation": "push",
            "statusBarAnimation": undefined,
            "statusBarHidden": undefined,
            "statusBarStyle": undefined,
            "style": {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
              "zIndex": undefined,
            },
            "swipeDirection": "horizontal",
            "transitionDuration": undefined,
          },
          "type": "RNSScreen",
        },
      ],
      "props": {
        "onFinishTransitioning": undefined,
        "style": {
          "flex": 1,
        },
      },
      "type": "RNSScreenStack",
    },
  ],
  "props": {
    "onInsetsChange": [Function],
    "style": [
      {
        "flex": 1,
      },
      undefined,
    ],
  },
  "type": "RNCSafeAreaProvider",
}
`;

exports[`NavigationFlow mostra telas de autenticação quando não logado 1`] = `
{
  "children": [
    {
      "children": null,
      "props": {
        "onInsetsChange": [Function],
        "style": [
          {
            "flex": 1,
          },
          {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          },
        ],
      },
      "type": "RNCSafeAreaProvider",
    },
    {
      "children": [
        {
          "children": [
            {
              "children": "[...]",
              "props": {
                "collapsable": false,
                "style": [
                  {
                    "flex": 1,
                  },
                  [
                    {
                      "backgroundColor": "rgb(242, 242, 242)",
                    },
                    undefined,
                  ],
                ],
              },
              "type": "RNSScreenContentWrapper",
            },
            {
              "children": "[...]",
              "props": {
                "backButtonDisplayMode": undefined,
                "backButtonInCustomView": false,
                "backTitle": undefined,
                "backTitleFontFamily": "System",
                "backTitleFontSize": undefined,
                "backTitleVisible": true,
                "backgroundColor": "rgb(255, 255, 255)",
                "blurEffect": undefined,
                "color": "rgb(0, 122, 255)",
                "direction": "ltr",
                "disableBackButtonMenu": false,
                "hidden": true,
                "hideBackButton": false,
                "hideShadow": undefined,
                "largeTitle": undefined,
                "largeTitleBackgroundColor": undefined,
                "largeTitleColor": undefined,
                "largeTitleFontFamily": "System",
                "largeTitleFontSize": undefined,
                "largeTitleFontWeight": "700",
                "largeTitleHideShadow": false,
                "pointerEvents": "box-none",
                "style": {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                  "position": "absolute",
                  "width": "100%",
                },
                "title": "Login",
                "titleColor": "rgb(28, 28, 30)",
                "titleFontFamily": "System",
                "titleFontSize": undefined,
                "titleFontWeight": "600",
                "topInsetEnabled": true,
                "translucent": false,
              },
              "type": "RNSScreenStackHeaderConfig",
            },
          ],
          "props": {
            "activityState": 2,
            "aria-hidden": false,
            "collapsable": false,
            "customAnimationOnSwipe": undefined,
            "fullScreenSwipeEnabled": undefined,
            "fullScreenSwipeShadowEnabled": true,
            "gestureEnabled": undefined,
            "gestureResponseDistance": {
              "bottom": -1,
              "end": -1,
              "start": -1,
              "top": -1,
            },
            "hasLargeHeader": false,
            "hideKeyboardOnSwipe": undefined,
            "homeIndicatorHidden": undefined,
            "nativeBackButtonDismissalEnabled": false,
            "navigationBarHidden": undefined,
            "onAppear": [Function],
            "onDisappear": [Function],
            "onDismissed": [Function],
            "onGestureCancel": [Function],
            "onHeaderBackButtonClicked": [Function],
            "onHeaderHeightChange": [Function],
            "onNativeDismissCancelled": [Function],
            "onSheetDetentChanged": [Function],
            "onTransitionProgress": [Function],
            "onWillAppear": [Function],
            "onWillDisappear": [Function],
            "preventNativeDismiss": undefined,
            "replaceAnimation": "push",
            "screenOrientation": undefined,
            "sheetAllowedDetents": [
              1,
            ],
            "sheetCornerRadius": -1,
            "sheetElevation": 24,
            "sheetExpandsWhenScrolledToEdge": true,
            "sheetGrabberVisible": false,
            "sheetInitialDetent": 0,
            "sheetInitialDetentIndex": 0,
            "sheetLargestUndimmedDetent": -1,
            "sheetLargestUndimmedDetentIndex": -1,
            "stackAnimation": undefined,
            "stackPresentation": "push",
            "statusBarAnimation": undefined,
            "statusBarHidden": undefined,
            "statusBarStyle": undefined,
            "style": {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
              "zIndex": undefined,
            },
            "swipeDirection": "horizontal",
            "transitionDuration": undefined,
          },
          "type": "RNSScreen",
        },
      ],
      "props": {
        "onFinishTransitioning": undefined,
        "style": {
          "flex": 1,
        },
      },
      "type": "RNSScreenStack",
    },
  ],
  "props": {
    "onInsetsChange": [Function],
    "style": [
      {
        "flex": 1,
      },
      undefined,
    ],
  },
  "type": "RNCSafeAreaProvider",
}
`;
